les/lespay/server/balance.go

```diff
@@ -0,0 +1,803 @@
+		// BalanceField contains the NodeBalance struct which implements nodePriority,
+		// allowing on-demand priority calculation and future priority estimation
+		BalanceField: setup.NewField("balance", reflect.TypeOf(&NodeBalance{})),
+	}
+}
+
+// Init sets the fields used by BalanceTracker as an input
+func (bts *BalanceTrackerSetup) Init(negBalanceKeyField, capacityField nodestate.Field) {
+	bts.negBalanceKeyField = negBalanceKeyField
+	bts.capacityField = capacityField
+}
+
+// BalanceTracker tracks positive and negative balances for connected nodes.
+// After negBalanceKeyField is set externally, a NodeBalance is created and previous
+// balance values are loaded from the database. Both balances are exponentially expired
+// values. Costs are deducted from the positive balance if present, otherwise added to
+// the negative balance. If the capacity is non-zero then a time cost is applied
+// continuously while individual request costs are applied immediately.
+// The two balances are translated into a single priority value that also depends
+// on the actual capacity.
+type BalanceTracker struct {
+	BalanceTrackerSetup
+	clock                      mclock.Clock
+	lock                       sync.Mutex
+	ns                         *nodestate.NodeStateMachine
+	ndb                        *nodeDB
+	posExp, negExp             utils.ValueExpirer
+	posExpTC, negExpTC         uint64
+	posThreshold, negThreshold uint64
+	totalAmount                utils.ExpiredValue
+	lastActiveBalanceUpdate    mclock.AbsTime
+	quit                       chan struct{}
+}
+
+// NewBalanceTracker creates a new BalanceTracker
+func NewBalanceTracker(ns *nodestate.NodeStateMachine, setup BalanceTrackerSetup, db ethdb.KeyValueStore, clock mclock.Clock, posExp, negExp utils.ValueExpirer) *BalanceTracker {
+	ndb := newNodeDB(db, clock)
+	bt := &BalanceTracker{
+		ns:                  ns,
+		BalanceTrackerSetup: setup,
+		ndb:                 ndb,
+		clock:               clock,
+		posExp:              posExp,
+		negExp:              negExp,
+		quit:                make(chan struct{}),
+	}
+	var start enode.ID
+	for {
+		ids := bt.ndb.getPosBalanceIDs(start, enode.ID{}, 1000)
+		var stop bool
+		l := len(ids)
+		if l == 1000 {
+			l--
+			start = ids[l]
+		} else {
+			stop = true
+		}
+		for i := 0; i < l; i++ {
+			bt.totalAmount.AddExp(bt.ndb.getOrNewBalance(ids[i].Bytes(), false))
+		}
+		if stop {
+			break
+		}
+	}
+
+	ns.SubscribeField(bt.capacityField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		n, _ := ns.GetField(node, bt.BalanceField).(*NodeBalance)
+		if n == nil {
+			log.Error("Capacity field changed while node field is missing")
+			return
+		}
+		newCap, _ := newValue.(uint64)
+		if newCap != 0 {
+			n.setCapacity(newCap)
+		} else {
+			n.deactivate(false)
+		}
+	})
+
+	ns.SubscribeField(bt.negBalanceKeyField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
```

Comment:
What we can retrieve from `enode.Node` is the IP announced by the remote node which might be false or missing. We can think about how to make the code nicer but I think we should indeed get this value from `p2p.Peer` and somehow pass it to `BalanceTracker`. Also, under certain circumstances (like testing) we might want to specify free IDs in a different way.

Reply:
We can of course change the name if `negBalanceKeyField` is a bad name. Maybe `freeIdField`? (I am always open to suggestions about naming things because I am generally not good at it)